feat(incr5): added processing of env variables

Kenny201 opened the pull request.

Commit:
feat(incr1): create server with handlers

Commit:
feat(incr1): add logic for create short url

- add urlgenerator package
- add storage package
- edit handlers for get, post methods

Commit:
feat(incr1): edit naming variable

Commit:
feat(incr1): delete field required

Commit:
feat(incr1): replaces redirect on set header location

Commit:
feat(incr1): delete arg request not used

Commit:
feat(incr1): edit handlers post

Commit:
feat(incr1): modified function create short url on private

Commit:
feat(incr1): rename methods handlers

Commit:
feat(incr1): update go version to 1.22.5

Commit:
feat(incr1): update routing and handlers

Commit:
feat(incr1): add type method into handeFunt

Commit:
feat(incr2): add testing for handlers
- add library testify
- add tests for http handlers

Commit:
feat(incr1): add http error if not found url

Commit:
feat(incr2): edit test for ddd pattern

Commit:
feat(incr1): rename methods and variables

Commit:
feat(incr1): added ddd pattern into project

Commit:
feat(incr3): add chi route framework
- edit handler with chi routes
- edit tests for chi route

Commit:
feat(incr4): added processing of command line arguments
- add config package with flags
- changed handlers to work with arguments

Commit:
feat(incr1): delete return error for put method

Commit:
feat(incr4): add package url_parser

Commit:
feat(incr5): added processing of env variables

Commit:
feat(incr1): edit error handling

Commit:
feat(incr2): edit error handling

Commit:
feat(incr5): delete not used fmt import

Commit:
feat(incr5): rename variables into setArgsFromEnv function

Commit:
feat(incr1): delete return error from put method repository

Commit:
feat(incr1): add value objects patterns

Commit:
feat(incr2): rename variables

Commit:
feat(incr4): refactor config comands for valueobjects pattern

Commit:
feat(incr5): refactor variables env for valueobjects pattern

Commit:
feat(incr1): add error handler into PostHandler method

Commit:
feat(incr4): rename baseURL to originalURL

Commit:
feat(incr1): return 404 if id not found

Commit:
feat(incr1): edit type for variable body into put service method

Commit:
feat(incr1): rename variable for Service

Commit:
feat(incr4): rename variable into valueobjects

Commit:
feat(incr1): rename file memory_repository into repository_memory

Commit:
feat(incr4): delete not used request from Put method

Commit:
feat(incr5): rename variable env

Commit:
feat(incr1): edit method GetAll into repository

Commit:
feat(incr3): corrected router methods chi

Commit:
feat(incr1): add response body close

vokinneberg commented on internal/app/shortener/service.go:
```diff
@@ -0,0 +1,85 @@
+	Sr shortener.Repository
+}
+
+func NewService(storages ...Storage) *Service {
```
То, как работает `NewService(...)` с `WithRepository(...)` немного сложно понять. Сам паттерн  WithRepository(...) ок, хотя по-идее `Storage` должна возвращать ссылку на сервис, для того чтобы можно было вызвать `WithRepository(...)` несколько раз -> `type Storage func(s *Service) *Service`.

Потом не совсем понятно, как предполагается использовать  коллекцию, которая передается, кака параметр: `storages ...Storage`, ведь последнее значение в коллекции всегда будет перезаписывать `s.Sr`?
Для пояснения было бы здорово либо добавить комментарии, объясняющие поток выполнения либо реализовать юнит тесты, проверяющие логику работы.

vokinneberg commented on internal/app/shortener/service.go:
```diff
@@ -0,0 +1,85 @@
+	}
+}
+
+func WithRepositoryMemory() Storage {
```
То же самое касается этой функции - мне кажется она лишняя, достаточно создать репозиторий там, где вызывается `NewService(...)` и передать его туда `(infra.NewRepositoryMemory()`.

vokinneberg commented on internal/app/shortener/service.go:
```diff
@@ -0,0 +1,85 @@
+		if key, ok := s.Sr.CheckExistsOriginalURL(url); ok {
+			body = key.ShortURL().ToString()
+		} else {
+			urlEntity := aggregate.NewURL(originalURL, shortURL)
```
По-моему строки 60:62 и строки 65:67 совершенно идентичные. Можно ли как то сократить эту логику?

vokinneberg commented on internal/app/shortener/service.go:
```diff
@@ -0,0 +1,85 @@
+			body = key.ShortURL().ToString()
+		} else {
+			urlEntity := aggregate.NewURL(originalURL, shortURL)
+			urlEntity, err = s.Sr.Put(urlEntity)
```
Если мы сократим логику ошибку будет удобно обрабатывать сразу.

vokinneberg commented on internal/app/shortener/service.go:
```diff
@@ -0,0 +1,85 @@
+	result, err := s.Sr.Get(url)
+
+	if err != nil {
+		return nil, err
```
Не обязательно, но можно добавить контекст к ошибке:
```suggestion
		return nil, fmt.Errorf("url not found %v", err)
```

vokinneberg commented on internal/domain/shortener/repository.go:
```diff
@@ -0,0 +1,12 @@
+	"github.com/Kenny201/go-yandex-shortener.git/internal/domain/shortener/aggregate"
+)
+
+type Repository interface {
```
👍 за реализацию интерфейса. Нет необходимости реализовывать интерфейс в отдельном файле, я рекомендовал бы интерфейс объявить там, где он используется, а именно в файле сервиса.

vokinneberg commented on internal/infra/repository_memory.go:
```diff
@@ -0,0 +1,45 @@
+)
+
+type RepositoryMemory struct {
+	urls map[string]*aggregate.URL
```
На будущее - стоит подумать о том, как хранилище будет себя вести в условиях большого количества конкурентных запросов, причем на чтение запросов будет кратно больше, чем на запись.

vokinneberg commented on internal/infra/repository_memory.go:
```diff
@@ -0,0 +1,45 @@
+	}
+}
+
+func (rm *RepositoryMemory) Get(id string) (*aggregate.URL, error) {
+	if _, ok := rm.urls[id]; !ok {
+		err := errors.New("url not found")
+		return nil, err
+	}
+
+	return rm.urls[id], nil
+}
```
Эту функцию можно немного оптимизировать, хотя доступ к элементу мапы по сложности O(1), все-таки хорошим тоном будет получать значение один раз:
```suggestion
	url, ok := rm.urls[id]
	if !ok {
		return nil, errors.New("url not found")
	}
	return url, nil
```

vokinneberg commented on internal/http/url_handler.go:
```diff
@@ -0,0 +1,62 @@
+}
+
+func (sh ShortenerHandler) PostHandler(w http.ResponseWriter, r *http.Request) {
+	body, err := io.ReadAll(r.Body)
```
Не стоит забывать закрывать тело запроса после чтения.

```suggestion
	body, err := io.ReadAll(r.Body)
	defer r.Body.Close()
```

vokinneberg commented on internal/http/url_handler_test.go:
```diff
@@ -0,0 +1,85 @@
+			ctype, "text/plain")
+	}
+
+	_, err = io.ReadAll(res.Body)
```
Не стоит забывать закрывать тело ответа даже в тестах.

```suggestion
	_, err = io.ReadAll(res.Body)
	defer res.Body.Close()
```

vokinneberg commented on internal/http/url_handler_test.go:
```diff
@@ -0,0 +1,85 @@
+	_, err = io.ReadAll(res.Body)
+
+	if err != nil {
+		t.Fatalf("could not read response:%v", err)
```
👍 за дополнительные сообщения в тестах!

vokinneberg commented on internal/http/url_handler_test.go:
```diff
@@ -0,0 +1,85 @@
+	"testing"
+)
+
+func TestPostHandler(t *testing.T) {
```
Тесты реализованы, это здорово. Мне кажется тут не хватает разнообразия сценариев. Попробуй реализовать табличные тесты и покрыть негативные сценарии.

vokinneberg commented on internal/domain/shortener/valueobject/original_url.go:
```diff
@@ -0,0 +1,67 @@
+	scheme, hp, err := ParseOriginalURL(originalURL)
+	var port int
+
+	if len(hp) == 2 {
+		port, err = strconv.Atoi(hp[1])
+	}
+
+	if err != nil {
+		return OriginalURL{}, err
+	}
```
Стоит перенести обработку ошибок как можно ближе к самому вызову.
```suggestion
	if len(hp) == 2 {
		port, err = strconv.Atoi(hp[1])
	         if err != nil {
		          return OriginalURL{}, err
	         }
	}
```

vokinneberg commented on internal/domain/shortener/valueobject/original_url.go:
```diff
@@ -0,0 +1,67 @@
+	return fmt.Sprintf("%s://%s", ou.scheme, ou.host)
+}
+
+func ParseOriginalURL(s string) (string, []string, error) {
```
Попробуй использовать пакет `net/url` для парсинга URL из стандартной библиотеки.

vokinneberg commented on internal/domain/shortener/valueobject/base_url.go:
```diff
@@ -0,0 +1,64 @@
+	return bu.port
+}
+
+func ParseBaseURL(s string) (string, []string, error) {
```
Попробуй использовать пакет  `net/url` для парсинга URL из стандартной библиотеки.

vokinneberg commented on internal/domain/shortener/valueobject/base_url.go:
```diff
@@ -0,0 +1,64 @@
+package valueobject
```
Мне кажется логика base_url и original_url во многом  дублируется. Подумай о том, как можно реализовать общий компонент.

vokinneberg reviewed (changes requested):
Поздравляю с завершением первого спринта и первым ревью! 🤗 Твой код хорошо структурирован, покрыт тестами, работает и проходит тесты. Отличная работа! 🚀

**Сильные стороны твоего PR:**

1. **Функциональность:** Код функционален и корректен! 🚀
2. **Структура сервиса:** По условию задания структура сервиса должна представлять собой набор пакетов и структур, разделённых на слои. Твой код неплохо структурирован и разбит на отдельные компоненты и пакеты. 👍
3. **Использование интерфейсов:** В твоём коде реализованы интерфейс  репозитория, что позволяет отделить конкретную реализацию от кода, который использует интерфейс. 👍
4. **Тестирование:** Реализованы тесты базовых сценариев. 👍

**На что я бы рекомендовал обратить внимание и что можно улучшить:**

1. **Структура сервиса:** По условию задания структура сервиса должна представлять собой набор пакетов и структур, разделённых на слои. В твоем коде уже есть компонент конфига, но я бы рекомендовал выделить отдельный слой для хранилища. 👍
2. **Сложность:** Мне показалось, что некоторые места в коде излишне переусложнены. Попробуй упростить компонент для работы с  URL, можно попробовать использовать пакет `net/url` из стандартной библиотеки. Так же можно упростить работу сервиса.
3. **Синхронизация доступа** **к хранилищу.** На будущее стоит подумать, как будет себя вести код при конкурентной обработке нескольких запросов и как можно синхронизировать доступ к хранилищу.
4. **Комментарии:** Было бы здорово, если бы ты добавил комментарии в коде там, где есть сложные места, требующие пояснения, а также для публичных методов.
5. **Тестирование:** Было бы здорово добавить больше тестов для негативных сценариев или разбить существующие тесты на тесты поменьше, которые проверяют один сценарий. Попробуй использовать табличные тесты.

Commit:
feat(incr5): refactor after review mentor

Commit:
feat(incr5): refactor after review mentor
- add flag tests
- delete method WithRepository()
- delete valueobject originalURL
- added table tests for handlers
- add comments
- added url string parsing using the net/http package
- small fix

Commit:
feat(incr5): refactor after review mentor
- add id not found test into TestGetByIDHandler

vokinneberg reviewed (approved):
Привет 👋, вижу ты реализовал улучшения по моим комментариям и мне кажется код стал более структурированный, его стало легче читать. Здорово! 👍   Поздравляю с завершением первого спринта и завершением первого ревью! 🤗 Отличная работа! 🚀

Commit:
feat(incr5): small fix
